libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -30,48 +30,86 @@
 
 #include <libsolutil/Common.h>
 
-using namespace std;
 using namespace solidity;
 using namespace solidity::frontend;
 using namespace solidity::langutil;
 
+using std::optional;
+using std::nullopt;
+using std::string;
+
 void ConstantEvaluator::endVisit(UnaryOperation const& _operation)
 {
-	auto sub = evaluatedValue(_operation.subExpression());
-	if (sub)
-		setValue(_operation, sub->unaryOperatorResult(_operation.getOperator()));
+	if (auto const sub = result(_operation.subExpression()); sub.has_value())
+	{
+		auto const res = sub.value().type->unaryOperatorResult(_operation.getOperator());
+		if (auto const rationalType = dynamic_cast<RationalNumberType const*>(res.get()))
+		{
+			auto const subType = sub.value().type;
+			if (subType && subType->category() == Type::Category::Integer)
```

Comment:
Yes, because that's the annotated type information. not the RationalNumberType that is storing the rational.

Reply:
I mean, maybe not for the unary operator case, hmm..

Reply:
You are saying you are modifying the annotation in this file so that you can process the constant? If yes, then this will produce a different type annotation than the type checker - I would like to avoid that.

Reply:
Ah, maybe I misunderstood. In any case: This code has to respect that actual integer type and not just the fact that it is non-fractional - we have to check if the result is larger than max or smaller than min and wrap or break accordingly.